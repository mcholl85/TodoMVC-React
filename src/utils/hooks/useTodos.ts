import { useMutation, useQuery, useQueryClient, UseQueryResult } from '@tanstack/react-query'
import { createTodo, deleteTodo, deleteTodos, fetchTodos, updatePartiallyTodo } from '../api/todos'

export interface TodoType {
  id: string
  title: string
  completed: boolean
}

export function useTodos(filter: string) {
  const queryClient = useQueryClient()
  const { data: todos }: UseQueryResult<TodoType[]> = useQuery({
    queryKey: ['todos'],
    queryFn: fetchTodos,
  })

  const { mutate: saveTodo } = useMutation({
    mutationFn: createTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const displayedTodos =
    todos?.filter((todo) => {
      switch (filter) {
        case 'active':
          return !todo.completed
        case 'completed':
          return todo.completed
        default:
          return true
      }
    }) || []

  const allTodosCompleted = displayedTodos?.every((todo) => todo.completed)

  const { mutate: removeTodo } = useMutation({
    mutationFn: deleteTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const { mutate: setTitle } = useMutation({
    mutationFn: updatePartiallyTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const { mutate: setCompleted } = useMutation({
    mutationFn: updatePartiallyTodo,
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const { mutate: clearCompleted } = useMutation({
    mutationFn: () => deleteTodos(true),
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['todos'] })
    },
  })

  const changeAllCompletedTodo = (bool: boolean) => {
    todos?.forEach((todo) => {
      if (todo.completed === bool) setCompleted({ id: todo.id, completed: bool })
    })
  }

  return {
    todos: displayedTodos,
    count: {
      all: todos?.length || 0,
      completed: todos?.filter((todo) => todo.completed).length || 0,
      active: todos?.filter((todo) => !todo.completed).length || 0,
    },
    allTodosCompleted,
    saveTodo,
    removeTodo,
    setTitle,
    setCompleted,
    clearCompleted,
    changeAllCompletedTodo,
  }
}
