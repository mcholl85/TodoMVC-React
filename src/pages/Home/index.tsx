import classNames from 'classnames'
import { useState } from 'react'
import CreateTodo from '../../components/CreateTodo'
import Footer from '../../components/Footer'
import Todo from '../../components/Todo'
import { useTodos } from '../../utils/hooks/useTodos'

type HomeProps = {
  filter: 'all' | 'active' | 'completed'
}

export default function Home({ filter }: HomeProps) {
  const {
    todos,
    count: { all, active, completed },
    allTodosCompleted,
    saveTodo,
    changeAllCompletedTodo,
    setCompleted,
    setTitle,
    removeTodo,
    clearCompleted,
  } = useTodos(filter)
  const [isEditing, setIsEditing] = useState('')

  return (
    <>
      <section className='todoapp'>
        <header className='header'>
          <h1>todos</h1>
          <CreateTodo saveTodo={saveTodo} />
        </header>
        <section className='main'>
          <input
            onChange={(e) => changeAllCompletedTodo(e.target.checked)}
            checked={allTodosCompleted}
            id='toggle-all'
            className='toggle-all'
            type='checkbox'
          />
          <label htmlFor='toggle-all'>Mark all as complete</label>
          <ul className='todo-list'>
            {todos?.map((todo) => (
              <li
                data-testid={`list-${todo.id}`}
                key={todo.id}
                onDoubleClick={() => setIsEditing(todo.id)}
                className={classNames(
                  { completed: todo.completed },
                  { editing: isEditing === todo.id },
                )}
              >
                <Todo
                  key={todo.id}
                  id={todo.id}
                  title={todo.title}
                  completed={todo.completed}
                  setCompleted={setCompleted}
                  setTitle={setTitle}
                  removeTodo={removeTodo}
                  isEditing={isEditing}
                  setIsEditing={setIsEditing}
                />
              </li>
            ))}
          </ul>
        </section>
        {all > 0 ? (
          <Footer
            clearCompleted={clearCompleted}
            nbOfActiveTodos={active}
            nbOfCompletedTodos={completed}
          />
        ) : (
          ''
        )}
      </section>
      <footer className='info'>
        <p>Double-click to edit a todo</p>
        <p>
          Created by <a href='http://todomvc.com'>Matthieu CHOLLET</a>
        </p>
        <p>
          Part of <a href='http://todomvc.com'>TodoMVC</a>
        </p>
      </footer>
    </>
  )
}
