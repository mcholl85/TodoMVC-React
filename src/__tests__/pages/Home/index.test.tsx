import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { MemoryRouter } from 'react-router'
import Home from '../../../pages/Home'
import { useTodos } from '../../../utils/hooks/useTodos'

const mockTodos = [
  {
    id: '627ccf8e-96cf-4607-80a5-22b01a35c7eb',
    title: 'test',
    completed: false,
  },
  {
    id: 'a4ffab12-209b-4007-acc5-9b21119d858d',
    title: 'hello',
    completed: true,
  },
  {
    id: '543aa59f-ed63-49d4-9d83-dbf38586ff46',
    title: 'world',
    completed: false,
  },
]

jest.mock('../../../utils/hooks/useTodos', () => {
  return {
    ...jest.requireActual('../../../utils/hooks/useTodos'),
    useTodos: jest.fn(),
  }
})

const mockUseTodos = useTodos as jest.Mock

describe('Page Home', () => {
  test('Matches DOM Snapshot', () => {
    mockUseTodos.mockImplementation(() => ({
      todos: [],
      count: { all: 0 },
      getFilteredTodos: jest.fn(),
    }))

    const { asFragment } = render(<Home filter='all' />)

    expect(asFragment()).toMatchSnapshot()
  })

  test('select all todos', () => {
    const mockChangeAllCompletedTodo = jest.fn()
    mockUseTodos.mockImplementation(() => ({
      todos: [],
      count: { all: 0 },
      changeAllCompletedTodo: mockChangeAllCompletedTodo,
      getFilteredTodos: jest.fn(),
    }))

    render(<Home filter='all' />)
    const checkbox = screen.getByRole('checkbox', { name: /Mark all as complete/i })

    expect(checkbox).not.toBeChecked()

    userEvent.click(checkbox)

    expect(mockChangeAllCompletedTodo).toHaveBeenCalledTimes(1)
    expect(mockChangeAllCompletedTodo).toHaveBeenNthCalledWith(1, true)
  })

  test('unselect all todos', () => {
    const mockChangeAllCompletedTodo = jest.fn()
    mockUseTodos.mockImplementation(() => ({
      todos: [],
      count: { all: 0 },
      allTodosCompleted: true,
      changeAllCompletedTodo: mockChangeAllCompletedTodo,
      getFilteredTodos: jest.fn(),
    }))

    render(<Home filter='all' />)
    const checkbox = screen.getByRole('checkbox', { name: /Mark all as complete/i })

    expect(checkbox).toBeChecked()

    userEvent.click(checkbox)

    expect(mockChangeAllCompletedTodo).toHaveBeenCalledTimes(1)
    expect(mockChangeAllCompletedTodo).toHaveBeenNthCalledWith(1, false)
  })

  test('display footer and todolist when there are todos', () => {
    mockUseTodos.mockImplementation(() => ({
      todos: mockTodos,
      count: { all: mockTodos.length },
      allTodosCompleted: true,
      getFilteredTodos: jest.fn(),
    }))

    render(
      <MemoryRouter initialEntries={['/']}>
        <Home filter='all' />
      </MemoryRouter>,
    )

    const footer = screen.queryByTestId('footer')

    expect(footer).toBeInTheDocument()
  })

  test('not display footer nor todolist where there are no todos', () => {
    mockUseTodos.mockImplementation(() => ({
      todos: [],
      count: { all: 0 },
      allTodosCompleted: true,
      getFilteredTodos: jest.fn(),
    }))

    render(
      <MemoryRouter initialEntries={['/']}>
        <Home filter='all' />
      </MemoryRouter>,
    )

    const footer = screen.queryByTestId('footer')

    expect(footer).not.toBeInTheDocument()
  })
})
