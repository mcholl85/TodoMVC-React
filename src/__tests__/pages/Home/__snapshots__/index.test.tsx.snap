// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Home Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <section
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
        value=""
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      />
    </section>
  </section>
  <footer
    class="info"
  >
    <p>
      Double-click to edit a todo
    </p>
    <p>
      Created by 
      <a
        href="http://todomvc.com"
      >
        Matthieu CHOLLET
      </a>
    </p>
    <p>
      Part of 
      <a
        href="http://todomvc.com"
      >
        TodoMVC
      </a>
    </p>
  </footer>
</DocumentFragment>
`;
